### PR TITLE
simplewallet: Add Unicode input_line [Ryo backport]

### PR DESCRIPTION
**Pull request by fireice-uk**
**Issue description** : https://github.com/ryo-currency/ryo-currency/pull/102
**Monero Issue** : #4351

### CODE REVIEW

File: src/common/command_line.cpp [hunk 1]
```diff
@@ -36,6 +36,14 @@
 #include "cryptonote_config.h"
 #include "string_tools.h"
 
+#ifdef WIN32
```
Comment (moneromooo-monero) [inline src/common/command_line.cpp]: Please use _WIN32 (apparently WIN32 is only defined if you've got the right include included already, so it's error prone)

File: src/common/command_line.cpp [hunk 2]
```diff
@@ -74,6 +82,42 @@ namespace command_line
...
+    ReadConsoleW(hConIn, buffer, sizeof(buffer)/sizeof(wchar_t)-1, &read, nullptr);
+    buffer[read] = 0;
+
+    SetConsoleMode(hConIn, oldMode);
+    CloseHandle(hConIn);
+  
+    int size_needed = WideCharToMultiByte(CP_UTF8, 0, buffer, -1, NULL, 0, NULL, NULL);
+    std::string buf(size_needed, '\0');
+    WideCharToMultiByte(CP_UTF8, 0, buffer, -1, &buf[0], size_needed, NULL, NULL);
+    buf.pop_back(); //size_needed includes null that we needed to have space for
```
Comment (moneromooo-monero) [inline src/common/command_line.cpp]: Can you move that inner code to util please, it seems generally useful. It also has nothing to do with command line. Without the trim seems best. Then you can have the input_line kept in simplewallet calling it and trimming. Also gets rid of all the diff spam due to namespaces below.

File: src/common/command_line.cpp [hunk 3]
```diff
@@ -74,6 +82,42 @@ namespace command_line
...
+    SetConsoleMode(hConIn, ENABLE_LINE_INPUT | ENABLE_ECHO_INPUT | ENABLE_PROCESSED_INPUT);
+
+    wchar_t buffer[1024];
+    DWORD read;
+
+    ReadConsoleW(hConIn, buffer, sizeof(buffer)/sizeof(wchar_t)-1, &read, nullptr);
+    buffer[read] = 0;
+
+    SetConsoleMode(hConIn, oldMode);
+    CloseHandle(hConIn);
```
Comment (moneromooo-monero) [inline src/common/command_line.cpp]: Any of those return error codes that need checking ?

Comment (fireice-uk) [reply src/common/command_line.cpp]: I subscribe to school that return codes should be checked if something can be done about it. Here the only thing available would be to throw.

File: src/common/command_line.cpp (hunk 2 above)
Comment (fireice-uk) [reply src/common/command_line.cpp]: What do you mean by _inner code_ -> UCS2 conversion? Or do you want the _read_line_ to act as a wrapper on win?

File: src/common/command_line.cpp (hunk 1 above)
Comment (fireice-uk) [reply src/common/command_line.cpp]: What you are thinking is MSVC defines (not that the code compiles there anyway), but sure.

Comment (moneromooo-monero) [reply src/common/command_line.cpp]: No idea about windows. I'm repeating what I was told by someone who does know, but maybe I'm assuming things. If you think that is wrong, explain why.

File: src/common/command_line.cpp (hunk 2 above)
Comment (moneromooo-monero) [reply src/common/command_line.cpp]: I mean the whole code except the trim, which is simplewallet specific, as it's not a canonical way to behave for a line reading function. So simplewallet would keep a function that calls the utils input function, then trims.

**Comment by fireice-uk** [comment]
Let me know if that's what you meant.

File: src/common/command_line.cpp (hunk 1 above)
Comment (fireice-uk) [reply src/common/command_line.cpp]: To put it in simple terms:

gcc defines `WIN32` and `_WIN32`
msvc defines `_WIN32`

You use `WIN32` (and I did out of laziness), however you suggestion is better practice.

**Comment by moneromooo-monero** [comment]
Yes, that is what I meant. Note that seeds are input via input_secure_line, not input_line, so that one will also have to be changed.

**Comment by fireice-uk** [comment]
Looks like something you merged after the base of this PR. Hold on, let me rebase then.

**Comment by fireice-uk** [comment]
Ok, looks like someone already did most of the work on that one for me, apart from failing to call the correct API (`ReadConsoleW`), that should be ez. 

One note the existing code. UTF8 characters can be up to 4 bytes long. This kind of thing makes people scream in agony:

`      char ch;
      r = (TRUE == ::ReadConsoleA(h_cin, &ch, 1, &read, NULL));
`

File: src/common/password.cpp [hunk 4]
```diff
@@ -56,8 +56,6 @@ namespace
 
   bool read_from_tty(epee::wipeable_string& pass, bool hide_input)
   {
-    static constexpr const char BACKSPACE = 8;
```
Comment (fireice-uk) [inline src/common/password.cpp]: I have no idea why someone chose to do this instead of using `\b`

**Comment by fireice-uk** [comment]
Ok. Done. The second commit adds proper handling of **multibyte** characters.

**Comment by moneromooo-monero** [comment]
Ah, my mistake. I had been assuming you'd actually tried it at least once. If not, then we'll wait some more for a Windows coder to show up and do it.

**Comment by fireice-uk** [comment]
That's fine - if you are short on manpower I can do it tomorrow evening. It is purely a matter of me having to power down and switch to Windows.

**Comment by moneromooo-monero** [comment]
OK, that's fine. It's not like we're in a hurry about this now.

**Comment by fireice-uk** [comment]
@moneromooo-monero Compiled it on win and fixed two minor build errors. I think that's everything from my end.

File: src/common/password.cpp (hunk 4 above)
Comment (hyc) [reply src/common/password.cpp]: Most likely for symmetry with the Unix routine (which used BACKSPACE = 127, ASCII DEL). I'd suggest preserving this line. Although I would have named them both DELETE, not BACKSPACE.

File: src/common/password.cpp [hunk 5]
```diff
@@ -67,32 +65,43 @@ namespace
...
-      {
-        pass.push_back(ch);
-      }
+      
+      char utf8_ch[8] = {0};
+      int len;
+      if((len = WideCharToMultiByte(CP_UTF8, 0, &ucs2_ch, 1, utf8_ch, sizeof(utf8_ch), NULL, NULL)) <= 0)
+        break;
+
+      chlen.push_back(len);
```
Comment (hyc) [inline src/common/password.cpp]: It's possible for the UTF8 char to push pass.size() over the max_password_size. Should check the length and break out of the loop here, before appending the new character.

I think it's bad practice to just silently return with whatever was read so far, but the existing code already does this. And I suppose the likelihood of anyone entering such a long password is low to begin with, so maybe not a serious problem.

File: src/common/util.cpp [hunk 6]
```diff
@@ -939,4 +939,32 @@ std::string get_nix_version_display_string()
...
     return newval;
   }
+  
+#ifdef _WIN32
+  std::string input_line_win()
+  {
+    HANDLE hConIn = CreateFileW(L"CONIN$", GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr, OPEN_EXISTING, 0, nullptr);
+    DWORD oldMode;
+
+    FlushConsoleInputBuffer(hConIn);
```
Comment (hyc) [inline src/common/util.cpp]: Why can't we just use STD_INPUT_HANDLE here?

The Unix version will work with input redirection. I don't see why we should break that for Windows.

File: src/simplewallet/simplewallet.cpp [hunk 7]
```diff
@@ -187,9 +146,10 @@ namespace
     std::cout << prompt;
 
     std::string buf;
+#ifdef _WIN32
+    buf = tools::input_line_win();
```
Comment (hyc) [inline src/simplewallet/simplewallet.cpp]: Probably should have emulated std::getline() semantics instead, e.g.
  `tools::getline(buf);`

**Comment by moneromooo-monero** [comment]
The buffer overflow should be fixed at least. If you want this in 0.13, fix it quick please.

**Comment by fireice-uk** [comment]
@moneromooo-monero  `reserve` is purely a memory optimisation - a C++ vector is dynamically sized (that's kind of the point). I can add a break if it is exceeded if you want though, it is two lines after-all =).

File: src/simplewallet/simplewallet.cpp (hunk 7 above)
Comment (fireice-uk) [reply src/simplewallet/simplewallet.cpp]: This paradigm is actually faster in C++11 onwards (copy elision).

**Comment by fireice-uk** [comment]
Let me know if that's what you meant. Whenever dealing with Unicode you should really delete all `size` variables. You need to be specific whether you mean bytes, characters or glyphs since the count of all three can be different.

File: src/common/password.cpp [hunk 8]
```diff
@@ -67,32 +66,46 @@ namespace
...
-      {
-        pass.push_back(ch);
-      }
+      
+      char utf8_ch[8] = {0};
+      int len;
+      if((len = WideCharToMultiByte(CP_UTF8, 0, &ucs2_ch, 1, utf8_ch, sizeof(utf8_ch), NULL, NULL)) <= 0)
+        break;
+
+      if(std::accumulate(chlen.begin(), chlen.end(), 0) + len >= tools::password_container::max_password_size)
```
Comment (hyc) [inline src/common/password.cpp]: `if(pass.size() + len >= tools::password_container::max_password_size)`

**Comment by hyc** [comment]
In this context we have to assume size means bytes.

**Comment by moneromooo-monero** [comment]
Could be, I didn't look at the code hyc pointed out. If there's no bug, then it's good to go. If it's just a misunderstanding, then it can be merged now.

File: src/common/password.cpp (hunk 8 above)
Comment (fireice-uk) [reply src/common/password.cpp]: Updated

**Pull request merged**